x/poe/keeper/keeper.go

```diff
@@ -117,6 +120,14 @@ func (k *Keeper) GetBondDenom(ctx sdk.Context) string {
 	return types.DefaultBondDenom
 }
 
+func (k *Keeper) UpdateValidatorVotes(validatorVotes []types2.VoteInfo) {
+	k.validatorVotes = validatorVotes
```

Comment:
Now that I think of it, I think this assignment is not enough, as slices are references to the underlying array. I think we need to use `copy` here, because we cannot guarantee the lifetime / validity of the `validatorVotes` data.

Reply:
Perhaps an integration test can be written, to check if copying the slice is really needed here.

Reply:
That, or a lifecycle analysis of the underlying array.

Reply:
After talks, we don't think this will be an issue. At least with the only current call to `ValidatorVotes` in the end blocker.

Leaving as Unresolved so that @alpe and @ethanfrey can take a look at it.

Reply:
This is not guaranteed to be thread safe. It is most likely but theoretically a concurrent query may be executed via contract in a simulation query.
https://github.com/tendermint/spec/blob/95cf253b6df623066ff7cd4074a94e7a3f147c7a/spec/abci/abci.md#overview